doc/design/block.md

```diff
@@ -0,0 +1,128 @@
+# Block design
+
+In computer programming, a block is a lexical structure of source code which is grouped together.
+In most programming languages, block is useful when define a function or some conditional statements such as `if` and `while`.
+
+In Paddle, we need a similar concept to support following scenes:
+
+- the basic unit of execution, just like the original `NetOp`, block can group multiple operators and make them behave just like a single operator. Unlike `NetOp`, a network should not be an operator, but a block can be treated as a operator.
+  - Block should have a `Exec` method, which will execute the operators grouped in the block.
```

Comment:
I think `Block::Exec` should also "realize" local variables, i.e., creating a Scope which is like a stack frame, and creating variables in it.  I think so because it is Block who knows about the local variable definitions, thus it is the only entity who can realize these variables.

Reply:
Currently, only one Scope is needed, and variables in Blocks are stored in this scope.

All the Variables will have unique names, so local variables in a Block will not affect the global namespace.

There is no difference whether a Block has a local scope or not.

Reply:
@wangkuiyi  the local variables inside a block will be used by the backward net too, so they should be able to be visited by Operator outside this block.